File: Data Structures & Algorithms/counting-bits/submission-2.java
class Solution {
    public int[] countBits(int n) {
        int[] res = new int[n+1];
        int num=1;
        for(int i=1;i<=n;i++){
            if((i&1)==1){
                res[i]=res[i>>1]+1;
            }
            else{
                if(i==(1<<num)){
                    res[i] = 1;
                    num++;
                }
                else{ res[i]=res[i>>1];}
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/subsets-ii/submission-2.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()

        def dfs(idx: int):

            if idx == len(nums):
                res.append(tmp[:])
                return
            
            tmp.append(nums[idx])
            dfs(idx+1)
            tmp.pop()
            while idx+1 < len(nums) and nums[idx]==nums[idx+1]:
                idx+=1
            
            dfs(idx+1)
        tmp = []
        dfs(0)
        return res